### PR TITLE
Implement full spec changes for Index/Range

### LINKED ISSUE #30620: Remove array/string special case indexer injection for Index & Range before shipping
Everything added in PR https://github.com/dotnet/roslyn/pull/30618 should be removed before shipping dev16.

### PR DESCRIPTION
**Pull request by agocke**
Reflects the changes described in https://github.com/dotnet/csharplang/pull/2214.

Range now looks optimistically for some special members with optional
parameters, but will fall back to the Range constructor if they aren't
present.

This change also removes all of the specialized codegen that was going
on with arrays and strings in favor of expected helpers or real members.

Fixes #30620

### CODE REVIEW

**Comment by jcouv** [comment]
Sounds like this fixes https://github.com/dotnet/roslyn/issues/31748. Can you add a test to verify/close?

**Comment by agocke** [comment]
@jcouv I'll do that in a follow-up PR

File: src/Compilers/Core/Portable/WellKnownMembers.cs [hunk 1]
```diff
@@ -3077,42 +3065,45 @@ static WellKnownMembers()
                      (byte)SignatureTypeCode.TypeHandle, (byte)SpecialType.System_Int32,
                      (byte)SignatureTypeCode.TypeHandle, (byte)SpecialType.System_Boolean,
 
-                 // System_Index__IsFromEnd
-                 (byte)(MemberFlags.Property),                                                                                                                  // Flags
+                 // System_Index__GetOffset
```
Comment (jcouv) [inline src/Compilers/Core/Portable/WellKnownMembers.cs]: >System_Index__GetOffset [](start = 20, length = 23)

I expected some tests to do `MakeTypeMissing(System_Index__GetOffset)` but couldn't find any. Do we have coverage for missing types?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -1974,7 +1974,43 @@ private BoundExpression BindRangeExpression(RangeExpressionSyntax node, Diagnost
 
             if (!rangeType.IsErrorType())
             {
-                symbolOpt = (MethodSymbol)GetWellKnownTypeMember(Compilation, WellKnownMember.System_Range__ctor, diagnostics, syntax: node);
+                // Depending on the available arguments to the range expression, there are four
+                // possible well-known members we could bind to. The constructor is always the
+                // fallback member, usable in any situation. However, if any of the other members
+                // are available and applicable, we will prefer that.
+
+                WellKnownMember? wmOpt = null;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: >wmOpt [](start = 33, length = 5)

nit: memberOpt?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs [hunk 3]
```diff
@@ -93,21 +74,27 @@ BoundExpression tryOptimizeOperand(BoundExpression operand)
...
 
-            // left.HasValue && right.HasValue
+            // if (left.HasValue && right.HasValue)
+            //    makeRange(left.GetValueOrDefault(), right.GetValueOrDefault())
             BoundExpression condition = null;
-            foreach (var operand in operands)
+            left = getIndexFromPossibleNullable(left);
+            right = getIndexFromPossibleNullable(right);
+            var rangeExpr = MakeRangeExpression(node.MethodOpt, left, right);
+            BoundExpression getIndexFromPossibleNullable(BoundExpression arg)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs]: local function in the middle of the method and without spacing is confusing for main flow of logic. Can it be moved at the end of the method?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs [hunk 4]
```diff
@@ -93,21 +74,27 @@ BoundExpression tryOptimizeOperand(BoundExpression operand)
...
 
-            ArrayBuilder<BoundExpression> sideeffects = ArrayBuilder<BoundExpression>.GetInstance();
-            ArrayBuilder<LocalSymbol> locals = ArrayBuilder<LocalSymbol>.GetInstance();
-            ArrayBuilder<BoundExpression> arguments = ArrayBuilder<BoundExpression>.GetInstance();
+            var sideeffects = ArrayBuilder<BoundExpression>.GetInstance();
+            var locals = ArrayBuilder<LocalSymbol>.GetInstance();
 
-            // left.HasValue && right.HasValue
+            // if (left.HasValue && right.HasValue)
+            //    makeRange(left.GetValueOrDefault(), right.GetValueOrDefault())
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs]: >makeRange(left.GetValueOrDefault(), right.GetValueOrDefault()) [](start = 18, length = 62)

This comment doesn't match other comments below:
`// left.HasValue && right.HasValue ? new Nullable(method(left.GetValueOrDefault(), right.GetValueOrDefault())) : default`

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs [hunk 5]
```diff
@@ -168,5 +148,42 @@ private BoundExpression LiftRangeExpression(BoundRangeExpression node, Immutable
...
+                case MethodKind.Constructor:
+                    // The constructor can always be used to construct a range, but if
+                    // any of the arguments are missing then we need to construct replacement
+                    // Indexes
+                    left = left ?? newIndexZero(fromEnd: false);
+                    right = right ?? newIndexZero(fromEnd: true);
+
+                    return F.New(constructionMethod, ImmutableArray.Create(left, right));
+
+                case MethodKind.Ordinary:
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs]: >case MethodKind.Ordinary: [](start = 16, length = 25)

comments for these different cases would be useful to illustrate scenarios

**Comment by jcouv** [comment]
>             var left = node.LeftOperand;

Not related to this PR, but the two operands are optional. The bound nodes should be named with `Opt`.

---
Refers to: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs:21 in c648f79. [](commit_id = c648f79afe03360147215575627bd18d92789dc9, deletion_comment = False)

File: src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs [hunk 6]
```diff
@@ -1960,5 +1549,40 @@ static string Print(Range? arg)
         }
     }
 }";
+
+        // The references we use for System.String do not have an indexer for
+        // System.Index and System.Range, so this wrapper mimics the behavior
+        private const string StringWithIndexers = @"
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs]: >StringWithIndexers [](start = 29, length = 18)

Are there tests of `Index` and `Range` with `string` without this wrapper class?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs (hunk 5 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs]: Thanks!

---
In reply to: [260988057](https://github.com/dotnet/roslyn/pull/33679#discussion_r260988057) [](ancestors = 260988057)

**Comment by jcouv** [comment]
>             var left = node.LeftOperand;

You updated BoundNodes.xml, but not the names in C# code. Won't compile.

---
In reply to: [468070924](https://github.com/dotnet/roslyn/pull/33679#issuecomment-468070924) [](ancestors = 468070924)

---
Refers to: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Range.cs:21 in c648f79. [](commit_id = c648f79afe03360147215575627bd18d92789dc9, deletion_comment = False)

**Pull request merged**